core/test/base/batch_multi_vector.cpp

```diff
@@ -0,0 +1,451 @@
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/base/batch_multi_vector.hpp>
+
+
+#include <gtest/gtest.h>
+
+
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/base/range.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+
+
+#include "core/test/utils.hpp"
+
+
+template <typename T>
+class BatchMultiVector : public ::testing::Test {
+protected:
+    using value_type = T;
+    using DenseMtx = gko::matrix::Dense<value_type>;
+    using size_type = gko::size_type;
+    BatchMultiVector()
+        : exec(gko::ReferenceExecutor::create()),
+          mtx(gko::batch_initialize<gko::BatchMultiVector<value_type>>(
+              {{{-1.0, 2.0, 3.0}, {-1.5, 2.5, 3.5}},
+               {{1.0, 2.5, 3.0}, {1.0, 2.0, 3.0}}},
+              exec))
+    {}
+
+
+    static void assert_equal_to_original_mtx(
+        const gko::BatchMultiVector<value_type>* m)
+    {
+        ASSERT_NE(m->get_const_values(), nullptr);
+        EXPECT_EQ(m->get_const_values()[0], value_type{-1.0});
+        ASSERT_EQ(m->get_num_batch_entries(), 2);
+        ASSERT_EQ(m->get_common_size(), gko::dim<2>(2, 3));
+        EXPECT_EQ(m->at(0, 0, 0), value_type{-1.0});
+        EXPECT_EQ(m->at(0, 0, 1), value_type{2.0});
+        EXPECT_EQ(m->at(0, 0, 2), value_type{3.0});
+        EXPECT_EQ(m->at(0, 1, 0), value_type{-1.5});
+        EXPECT_EQ(m->at(0, 1, 1), value_type{2.5});
+        ASSERT_EQ(m->at(0, 1, 2), value_type{3.5});
+        EXPECT_EQ(m->at(1, 0, 0), value_type{1.0});
+        EXPECT_EQ(m->at(1, 0, 1), value_type{2.5});
+        EXPECT_EQ(m->at(1, 0, 2), value_type{3.0});
+        EXPECT_EQ(m->at(1, 1, 0), value_type{1.0});
+        EXPECT_EQ(m->at(1, 1, 1), value_type{2.0});
+        ASSERT_EQ(m->at(1, 1, 2), value_type{3.0});
+    }
+
+    static void assert_empty(gko::BatchMultiVector<value_type>* m)
+    {
+        ASSERT_EQ(m->get_num_batch_entries(), 0);
+        ASSERT_EQ(m->get_common_size(), gko::dim<2>{});
+        ASSERT_EQ(m->get_const_values(), nullptr);
+    }
+
+    std::shared_ptr<const gko::Executor> exec;
+    std::unique_ptr<gko::BatchMultiVector<value_type>> mtx;
+};
+
+TYPED_TEST_SUITE(BatchMultiVector, gko::test::ValueTypes);
+
+
+TYPED_TEST(BatchMultiVector, CanBeEmpty)
+{
+    auto empty = gko::BatchMultiVector<TypeParam>::create(this->exec);
+    this->assert_empty(empty.get());
+}
+
+
+TYPED_TEST(BatchMultiVector, KnowsItsSizeAndValues)
+{
+    ASSERT_NE(this->mtx->get_const_values(), nullptr);
+    this->assert_equal_to_original_mtx(this->mtx.get());
+}
+
+
+TYPED_TEST(BatchMultiVector, CanGetValuesForEntry)
+{
+    using value_type = typename TestFixture::value_type;
+
+    ASSERT_EQ(this->mtx->get_values_for_entry(1)[0], value_type{1.0});
+}
+
+
+TYPED_TEST(BatchMultiVector, CanBeCopied)
```

Comment:
I think this and `CanBeMoved` `CanBeCloned` should rather be in the `test/...` test, because they depend on the executor.

Reply:
We want to test the core functionality here, so tests using reference executor should be sufficient I think.